File: leetcode/0024-swap-nodes-in-pairs/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
    if head is None:
        return head

    prev = None
    first = head
    second = head.next

    while second:
        next_node = second.next

        first.next = next_node
        second.next = first

        if prev is not None:
            prev.next = second
        else:
            head = second

        # Increment the pointers
        prev = first
        first = first.next

        if first is None:
            break

        second = first.next

    return head

File: leetcode/0024-swap-nodes-in-pairs/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs(head: ListNode | null): ListNode | null {
  if (!head) {
    return head;
  }

  let prev: ListNode | null = null;
  let first: ListNode | null = head;
  let second: ListNode | null = head.next;

  while (second) {
    const nextNode: ListNode | null = second.next;

    first.next = nextNode;
    second.next = first;

    if (prev) {
      prev.next = second;
    } else {
      head = second;
    }

    prev = first;
    first = first.next;

    if (!first) {
      break;
    }

    second = first.next;
  }

  return head;
}
